device/Microsoft.Azure.Devices.Client/ModuleAuthenticationWithSakRefresh.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Client.Extensions;
+using System;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    // Implementing SAS Token refresh based on a SharedAccessKey (SAK).
+    internal class ModuleAuthenticationWithSakRefresh : ModuleAuthenticationWithTokenRefresh
+    {
+        private IotHubConnectionString _connectionString;
+
+        public ModuleAuthenticationWithSakRefresh(
+            string deviceId,
+            string moduleId,
+            IotHubConnectionString connectionString) : base(deviceId, moduleId)
+        {
+            _connectionString = connectionString;
+        }
+
+        protected override Task<string> SafeCreateNewToken(string iotHub, int suggestedTimeToLive)
```

Comment:
Nit: to avoid code duplication you could extract the common implementation into the abstract (similar to how you did with Populate).

Reply:
You know I thought of that, but in this case it is not very clear how to do that, while keeping the code logically correct. Refactoring the code to the base will leave the code in patches (you won't be able to look at the code an understand what is going on). Maybe you have a better idea, but I will leave it like this for now.